File: R/discan.R
######################################################################
#
# discan.R
#
# copyright (c) 2001-5, Karl W Broman, Johns Hopkins University
# last modified Apr, 2005
# first written Oct, 2001
# Licensed under the GNU General Public License version 2 (June, 1991)
# 
# Part of the R/qtl package
# Contains: discan
#
######################################################################

######################################################################
#
# discan: scan genome, calculating LOD scores with single QTL model
#         for a dichotomous trait 
#
######################################################################

discan <-
function(cross, pheno, method=c("em","mr"),
         addcovar=NULL, intcovar=NULL, maxit=4000, tol=1e-4,
         verbose=FALSE)
{
  method <- match.arg(method)

  n.ind <- nind(cross)
  n.chr <- nchr(cross)
  type <- class(cross)[1]
  if(is.null(addcovar)) {
    n.addcovar <- 0
    addcovar <- 0
  }
  else n.addcovar <- ncol(addcovar)
  if(is.null(intcovar)) {
    n.intcovar <- 0
    intcovar <- 0
  }
  else n.intcovar <- ncol(intcovar)

  if(method=="mr" && n.addcovar+n.intcovar>0)  {
    warning("Covariates ignored with method=\"mr\"; use \"em\" instead")
    n.addcovar <- n.intcovar <- addcovar <- intcovar <- 0
  }

  u <- unique(pheno)
  if(any(u != 0 & u != 1))
    stop("Phenotypes must be either 0 or 1.")

  # get null log liklihood
  if(n.addcovar > 0)
    nullfit <- glm(pheno ~ addcovar, family=binomial(link=logit))
  else
    nullfit <- glm(pheno ~ 1, family=binomial(link=logit))
  fitted <- nullfit$fitted
  nullcoef <- nullfit$coef
  llik0 <- sum(pheno*log10(fitted) + (1-pheno)*log10(1-fitted))

  results <- NULL

  # calculate genotype probabilities one chromosome at a time
  for(i in 1:n.chr) {

    chrtype <- class(cross$geno[[i]])
    if(chrtype=="X") {
      sexpgm <- getsex(cross)
      ac <- revisecovar(sexpgm,addcovar)
      n.ac <- ifelse(is.null(ac),0,ncol(ac))
      ic <- revisecovar(sexpgm,intcovar)
      n.ic <- ifelse(is.null(ic),0,ncol(ic))
    }
    else {
      sexpgm <- NULL
      ac <- addcovar
      n.ac <- n.addcovar
      ic <- intcovar
      n.ic <- n.intcovar
    }

    # get genotype names
    gen.names <- getgenonames(type,chrtype,"full",sexpgm)
    n.gen <- length(gen.names)

    # pull out genotype data (mr)
    # or genotype probabilities (em)
    if(method == "mr") {
      newgeno <- cross$geno[[i]]$data
      newgeno[is.na(newgeno)] <- 0 

      # discard partially informative genotypes
      if(type=="f2" || type=="f2ss") newgeno[newgeno>3] <- 0
      if(type=="4way") newgeno[newgeno>4] <- 0

      # revise X chromosome genotypes
      if(chrtype=="X" && (type=="bc" || type=="f2" || type=="f2ss"))
         newgeno <- reviseXdata(type, "full", sexpgm, geno=newgeno)

      n.pos <- ncol(newgeno)
      map <- cross$geno[[i]]$map
      if(is.matrix(map)) map <- map[1,]

      z <- .C("R_discan_mr",
              as.integer(n.ind),         # number of individuals
              as.integer(n.pos),         # number of markers
              as.integer(n.gen),         # number of possible genotypes
              as.integer(newgeno),       # genotype data
              as.integer(pheno),          # phenotype data
              result=as.double(rep(0,n.pos*(n.gen+1))),
              PACKAGE="qtl")

    }
    else {
      if(is.na(match("prob",names(cross$geno[[i]])))) { # need to run calc.genoprob
        warning("First running calc.genoprob.")
        cross <- calc.genoprob(cross)
      }
      genoprob <- cross$geno[[i]]$prob
      n.pos <- ncol(genoprob)

      # revise X chromosome genotypes
      if(chrtype=="X" && (type=="bc" || type=="f2" || type=="f2ss"))
         genoprob <- reviseXdata(type, "full", sexpgm, prob=genoprob)

      map <- create.map(cross$geno[[i]]$map,
                        attr(cross$geno[[i]]$prob,"step"),
                        attr(cross$geno[[i]]$prob,"off.end"))
      if(is.matrix(map)) map <- map[1,]

      if(n.ac + n.ic > 0) {

        start <- rep(nullcoef[1],n.gen)
        if(n.ac > 0)
          start <- c(start, nullcoef[-1])
        if(n.ic > 0)
          start <- c(start, rep(0, n.ic*(n.gen-1)))

        z <- .C("R_discan_covar",
                as.integer(n.ind),         # number of individuals
                as.integer(n.pos),         # number of markers
                as.integer(n.gen),         # number of possible genotypes
                as.double(genoprob),       # genotype probabilities
                as.double(ac),
                as.integer(n.ac),
                as.double(ic),
                as.integer(n.ic),
                as.integer(pheno),          # phenotype data
                as.double(start),
                result=as.double(rep(0,n.pos)),
                as.integer(maxit),
                as.double(tol),
                as.integer(verbose),
                PACKAGE="qtl")
      }
      else {
        z <- .C("R_discan_im",
                as.integer(n.ind),         # number of individuals
                as.integer(n.pos),         # number of markers
                as.integer(n.gen),         # number of possible genotypes
                as.double(genoprob),       # genotype probabilities
                as.integer(pheno),          # phenotype data
                result=as.double(rep(0,n.pos*(n.gen+1))),
                as.integer(maxit),
                as.double(tol),
                PACKAGE="qtl")
      }

    }
    z <- matrix(z$result,nrow=n.pos)

    if(method == "em") z[,1] <- z[,1] - llik0
    z[is.na(z[,1]),1] <- 0

    if(n.ac + n.ic > 0)
      colnames(z) <- c("lod")
    else
      colnames(z) <- c("lod",gen.names)
      
    w <- names(map)
    o <- grep("^loc\-*[0-9]+",w)
    if(length(o) > 0) # inter-marker locations cited as "c*.loc*"
      w[o] <- paste("c",names(cross$geno)[i],".",w[o],sep="")
    rownames(z) <- w
    
    z <- as.data.frame(z)
    z <- cbind(chr=rep(names(cross$geno)[i],length(map)), pos=map, z)
    rownames(z) <- w

    # get null log10 likelihood for the X chromosome
    adjustX <- FALSE
    if(chrtype=="X") {

      # determine which covariates belong in null hypothesis
      temp <- scanoneXnull(type, sexpgm)
      adjustX <- temp$adjustX
      dfX <- temp$dfX
      sexpgmcovar <- temp$sexpgmcovar
      sexpgmcovar.alt <- temp$sexpgmcovar.alt      

      if(adjustX) { # get LOD-score adjustment
        if(n.ac > 0) 
          nullfit <- glm(pheno ~ ac+sexpgmcovar,
                         family=binomial(link=logit))
        else 
          nullfit <- glm(pheno ~ sexpgmcovar,
                         family=binomial(link=logit))
        fitted <- nullfit$fitted
        llik0X <- sum(pheno*log10(fitted) + (1-pheno)*log10(1-fitted))

        # adjust LOD curve
        z[,3] <- z[,3] - (llik0X - llik0)
      }
    } 

    # if different number of columns from other chromosomes,
    #     expand to match
    if(!is.null(results) && ncol(z) != ncol(results)) {
      cnz <- colnames(z)
      cnr <- colnames(results)
      wh.zr <- match(cnz,cnr)
      wh.rz <- match(cnr,cnz)
      if(all(!is.na(wh.rz))) {
        newresults <- data.frame(matrix(NA,nrow=nrow(results),ncol=ncol(z)))
        dimnames(newresults) <- list(rownames(results), cnz)
        newresults[,cnr] <- results
        results <- newresults
        for(i in 2:ncol(results))
          if(is.factor(results[,i])) results[,i] <- as.numeric(results[,i])
      }
      else if(all(!is.na(wh.zr))) {
        newz <- data.frame(matrix(NA,nrow=nrow(z),ncol=ncol(results)))
        dimnames(newz) <- list(rownames(z), cnr)
        newz[,cnz] <- z
        z <- newz
        for(i in 2:ncol(z))
          if(is.factor(z[,i])) z[,i] <- as.numeric(z[,i])
      }
      else {
        newnames <- c(cnr, cnz[is.na(wh.zr)])

        newresults <- data.frame(matrix(NA,nrow=nrow(results),ncol=length(newnames)))
        dimnames(newresults) <- list(rownames(results), newnames)
        newresults[,cnr] <- results
        results <- newresults
        for(i in 2:ncol(results))
          if(is.factor(results[,i])) results[,i] <- as.numeric(results[,i])
        
        newz <- data.frame(matrix(NA,nrow=nrow(z),ncol=length(newnames)))
        dimnames(newz) <- list(rownames(z), newnames)
        newz[,cnz] <- z
        z <- newz
        for(i in 2:ncol(z))
          if(is.factor(z[,i])) z[,i] <- as.numeric(z[,i])
      }
    }

    results <- rbind(results, z)
  } # loop over chromosomes

  if(ncol(results) > 3) {
    # sort the later columns
    neworder <- c(colnames(results)[1:3],sort(colnames(results)[-(1:3)]))
    results <- results[,neworder]
  }

  class(results) <- c("scanone","data.frame")
  attr(results,"method") <- method
  attr(results,"type") <- type
  attr(results,"model") <- "binary"
  attr(results,"null.log10.lik") <- llik0
  if(adjustX) 
    attr(results,"null.log10.lik.X") <- llik0X
  results
}

# end of discan.R

File: src/discan.c
/**********************************************************************
 * 
 * discan.c
 *
 * copyright (c) 2001, Karl W Broman, Johns Hopkins University
 *
 * last modified Oct, 2001
 * first written Oct, 2001
 *
 * Licensed under the GNU General Public License version 2 (June, 1991)
 *
 * C functions for the R/qtl package
 *
 * These functions are for performing a genome scan with a 
 * single QTL model
 *
 * Contains: R_discan_mr, discan_mr,
 *           R_discan_im, discan_im
 *  
 **********************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>
#include "util.h"
#include "discan.h"

/**********************************************************************
 * 
 * R_discan_mr
 *
 * Wrapper for call from R; reorganizes genotype and result matrix
 * and calls discan_mr.
 * 
 **********************************************************************/

void R_discan_mr(int *n_ind, int *n_pos, int *n_gen,
		    int *geno, double *pheno, double *result)
{
  int **Geno;
  double **Result;

  reorg_geno(*n_ind, *n_pos, geno, &Geno);
  reorg_errlod(*n_pos, *n_gen+1, result, &Result);

  discan_mr(*n_ind, *n_pos, *n_gen, Geno, pheno, Result);
}

/**********************************************************************
 * 
 * R_discan_im
 *
 * Wrapper for call from R; reorganizes genotype prob and result matrix
 * and calls discan_im.
 * 
 **********************************************************************/

void R_discan_im(int *n_ind, int *n_pos, int *n_gen, 
		 double *genoprob, double *pheno, double *result, 
		 int *maxit, double *tol)
{
  double ***Genoprob, **Result, **work;

  reorg_genoprob(*n_ind, *n_pos, *n_gen, genoprob, &Genoprob);
  reorg_errlod(*n_pos, *n_gen+1, result, &Result);
  allocate_dmatrix(3, *n_gen, &work);

  discan_im(*n_ind, *n_pos, *n_gen, Genoprob, pheno, Result, 
	    *maxit, *tol, work);

}

/**********************************************************************
 * 
 * discan_mr
 *
 * Performs genotype scan using marker regression for a dichotomous trait
 * 
 * n_ind        Number of individuals
 *
 * n_pos        Number of marker positions
 *
 * n_gen        Number of different genotypes
 *
 * Geno         Genotype matrix
 *
 * pheno        Phenotype data, as a vector
 *
 * Result       Result matrix of size [n_pos x (n_gen+1)]; upon return, 
 *              the first column contains the RSS and the rest contain
 *              estimated genotype-specific probabilities
 *
 **********************************************************************/

void discan_mr(int n_ind, int n_pos, int n_gen, int **Geno, 
		  double *pheno, double **Result)
{
  int i, j, k, n, tp, *ng, *np;

  /* number of individuals in each genotype group */
  ng = (int *)R_alloc(n_gen, sizeof(int));

  /* number of individuals with phenotype=1 in each genotype group */
  np = (int *)R_alloc(n_gen, sizeof(int));

  for(i=0; i<n_pos; i++) {
    Result[0][i] = 0.0; n=tp=0; 
    for(j=0; j< n_gen; j++) {
      ng[j] = np[j] = 0;

      for(k=0; k<n_ind; k++) {
	if(Geno[i][k] == j+1) {
	  if(pheno[k]) {
	    np[j]++;
	    tp++;
	  }
	  n++; ng[j]++;
	}
      }
      
      if(ng[j] > 0)
	Result[j+1][i] = (double)np[j]/(double)ng[j];
      else /* no individuals with this genotype */
	Result[j+1][i] = NA_REAL;
    } /* loop over genotype groups */

    /* calculate LOD score */
    for(j=0; j<n_gen; j++) {
      if(np[j] > 0 && np[j] < ng[j]) 
	Result[0][i] += ((double)np[j]*log10(Result[j+1][i]) +
			 (double)(ng[j]-np[j])*log10(1.0-Result[j+1][i]));
    }
    if(tp > 0 && tp < n) 
      Result[0][i] -= ((double)tp*log10((double)tp/(double)n) +
		       (double)(n-tp)*log10((double)(n-tp)/(double)n));

  } /* loop over marker positions */
}

/**********************************************************************
 * 
 * discan_im
 *
 * Performs genotype scan using interval mapping.  (The multipoint
 * genotype probabilities have already been calculated in 
 * calc.genoprob)
 * 
 * n_ind        Number of individuals
 *
 * n_pos        Number of marker positions
 *
 * n_gen        Number of different genotypes
 *
 * Genoprob     Array of conditional genotype probabilities
 *
 * pheno        Phenotype data, as a vector
 *
 * Result       Result matrix of size [n_pos x (n_gen+1)]; upon return, 
 *              the first column contains the log10 likelihood and the 
 *              rest contain estimated genotype-specific probabilities
 *
 * maxit        Maximum number of iterations in the EM algorithm
 *
 * tol          Tolerance for determining convergence in EM
 *
 * work         Workspace of dimension 3 x n_gen
 *
 **********************************************************************/

void discan_im(int n_ind, int n_pos, int n_gen, double ***Genoprob,
	       double *pheno, double **Result, 
	       int maxit, double tol, double **work)
{
  int i, j, k, s, flag=0;
  double sw;

  for(i=0; i<n_pos; i++) { /* loop over marker positions */

    /* initiate EM */
    for(k=0; k<n_gen; k++) {
      Result[k+1][i] = sw=0.0;
      for(j=0; j<n_ind; j++) {
	sw += Genoprob[k][i][j]; /* k=gen, i=pos, j=ind */
	if(pheno[j]) Result[k+1][i] += Genoprob[k][i][j];
      }
      Result[k+1][i] /= sw;
    }

    for(s=0; s < maxit; s++) { /* EM iterations */
    
      /* copy over current estimates */
      for(k=0; k<n_gen; k++) {
	work[0][k] = Result[k+1][i]; 
	Result[k+1][i] = work[1][k] = 0.0;
      }
      
      for(j=0; j<n_ind; j++) { /* loop over individuals */
	/* E-step */
	sw = 0.0;
	for(k=0; k<n_gen; k++) {
	  work[2][k] = Genoprob[k][i][j];
	  if(pheno[j]) work[2][k] *= work[0][k];
	  else work[2][k] *= (1.0-work[0][k]);
	  sw += work[2][k];
	}
	for(k=0; k<n_gen; k++) work[2][k] /= sw;

	/* M-step */
	for(k=0; k<n_gen; k++) {
	  work[1][k] += work[2][k];
	  if(pheno[j]) Result[k+1][i] += work[2][k];
	}
      }
      
      /* complete M-step */
      for(k=0; k<n_gen; k++) 
	Result[k+1][i] /= work[1][k];

      /* check for convergence */
      flag = 0;
      for(k=0; k<n_gen; k++) {
	if(fabs(Result[k+1][i] - work[0][k]) > tol*(fabs(work[0][k])+tol*100.0)) {
	  flag = 1;
	  break;
	}
      }

      if(!flag) break;
    } /* end of EM iterations */

    if(flag) warning("Didn't converge!\n");

    /* calculate log lik */
    Result[0][i] = 0.0;
    for(j=0; j<n_ind; j++) {
      sw = 0.0;
      if(pheno[j]) 
	for(k=0; k<n_gen; k++) sw += Genoprob[k][i][j] * Result[k+1][i];
      else
	for(k=0; k<n_gen; k++) sw += Genoprob[k][i][j] * (1.0-Result[k+1][i]);
      Result[0][i] += log10(sw);
    }

  } /* end loop over marker positions */
}



/* end of discan.c */

